content/client-lib-development-guide/features.textile

```diff
@@ -318,22 +324,25 @@ h3(#realtime-connection). Connection
-**** @(RTN15c3)@ @CONNECTED@ @ProtocolMessage@ with a new @connectionId@, and an error in @error@. In this case, a new connection has been established, the resume has failed, the channels are no longer attached, and the error indicates the resume problem which should be emitted. The client library should immediately detach all channels, fail any queued messages on those channels, and set the @Channel#errorReason@ on each detached @Channel@. Additionally, the internal @msgSerial@ counter is reset so that the first message published to Ably will contain a @msgSerial@ value of @0@
-**** @(RTN15c4)@ @ERROR@ @ProtocolMessage@ indicating a fatal error in the connection. The server will close the transport immediately after. If the @ERROR@ is non-recoverable, the client will move to the @FAILED@ state triggering all attached channels to move to the @FAILED@ state as well
+**** @(RTN15c3)@ @CONNECTED@ @ProtocolMessage@ with a new @connectionId@, and an error in @error@. In this case, a new connection has been established, the resume was unsuccessful, the channels are no longer attached, and the emitted error indicates the cause of the unsuccessful resume. The client library should initiate an attach for channels that are in the @SUSPENDED@ state. For all channels in the @ATTACHING@ or @ATTACHED@ state, the client library should fail any previously queued messages for that channel and initiate a new attach i.e. a new @ATTACH@ @ProtocolMessage@ must be sent for each channel. An @ERROR@ will most likely be emitted when the channels become attached automatically due to a loss of continuity, see "@RTL12@":#RTL12. Finally, the internal @msgSerial@ counter is reset so that the first message published to Ably will contain a @msgSerial@ value of @0@
+**** @(RTN15c4)@ @ERROR@ @ProtocolMessage@ indicating a fatal error in the connection. The server will close the transport immediately after. The client should move to the @FAILED@ state triggering all attached channels to move to the @FAILED@ state as well. Additionally the @Connection#errorReason@ will be set should be set with the error received from Ably
+**** @(RTN15c5)@ @DISCONNECTED@ @ProtocolMessage@ indicating a recoverable error in the connection, such as a token token error. The process described in "RTN15h":RTN15h must be followed for a connection being resumed with a token error
```

Comment:
> There is fundamentally a difference:
> ...
> An ERROR is sent when no connection exists

I would say that an `ERROR` is sent when the system is not able to associate the connection attempt with an existing connection. For example, attempting to resume with expired credentials. It doesn't mean that the connection is definitely in error, but the connection attempt did not get as far as being associated with any existing connection.

> A DISCONNECTED is sent on existing connection

I think you mean sent on an existing transport ... but @SimonWoolf's suggestion last night was that you could indeed get a `DISCONNECTED` response even on a transport that was never connected, if the system wishes to communicate that the connection is not failed, only the transport or connection attempt.

So the `ERROR` case, received in response to a connection attempt with, should be the only situation where you need to look at the specific error code to know whether or not you can still expect the connection to exist and be recoverable.